src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+    dispatch::DispatchResult, ensure, traits::Randomness,
+};
+use system::ensure_signed;
+
+// Encoding library
+use codec::Encode;
+
+pub trait Trait: system::Trait + balances::Trait + contracts::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    // Something that provides randomness generation in the runtime.
+    type Randomness: Randomness<Self::Hash>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        // No owner set for token
+        TokenNoOwner,
+
+        // Got an overflow after adding
+        Overflow,
+
+        // Got an underflow after subing
+        Underflow,
+    }
+}
+
+decl_event!(
+    pub enum Event<T> 
+        where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash {
+
+        // Token transfer event
+        Transfer(Option<AccountId>, Option<AccountId>, Hash),
+
+        // One token approved to an account
+        Approval(AccountId, AccountId, Hash),
+
+        // All tokens owned are approved to other account
+        ApprovalForAll(AccountId, AccountId, bool),
+    }
+);
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ERC721 {
+        // Token count owned by an account
+        pub OwnedTokensCount get(balance_of): map T::AccountId => u64;
+
+        // Token owner's account id
+        pub TokenOwner get(owner_of): map T::Hash => Option<T::AccountId>;
+
+        // Token delegate's account id
+        pub TokenApprovals get(get_approved): map T::Hash => Option<T::AccountId>;
+
+        // Account deledate's account id
+        pub OperatorApprovals get(is_approved_for_all): map (T::AccountId, T::AccountId) => bool;
+
+        // Total count of minted token
+        pub TotalSupply get(total_supply): u64;
+
+        // Map of token index to token id
+        pub AllTokens get(token_by_index): map u64 => T::Hash;
+
+        // Last token index
+        pub AllTokensIndex: map T::Hash => u64;
+
+        // Map of account and index to token id
+        pub OwnedTokens get(token_of_owner_by_index): map (T::AccountId, u64) => T::Hash;
+
+        // Token id to index
+        pub OwnedTokensIndex: map T::Hash => u64;
+
+        // Value use to generate random value
+        pub Nonce: u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin  {
+        fn deposit_event() = default;
+
+        // Approve a token to an account
+        fn approve(origin, to: T::AccountId, token_id: T::Hash) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let owner = match Self::owner_of(token_id) {
+                Some(c) => c,
+                None => return Err(Error::<T>::TokenNoOwner.into()),
+            };
+
+            ensure!(to != owner, "Owner is implicitly approved");
+            ensure!(sender == owner || Self::is_approved_for_all((owner.clone(), sender.clone())), "You are not allowed to approve for this token");
+
```

Comment:
I think the checking owner_of can cover such case. For not minted token, the owner will return None. The map to Option is a very annoying thing for pallet storage. Sometimes, we check the exists, sometimes, we check the returned value is None. As I know, some team suggest not use map to Option since it bring the ambiguity. If any suggested best practice from Centrifuge team?

Reply:
Makes sense, thanks! I think the Option is fine, it's also used all over substrate's code.